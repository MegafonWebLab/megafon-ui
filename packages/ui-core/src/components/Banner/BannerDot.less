@import '~styles/base.less';
@b: e('.mfui-beta-banner-dot');

@{b} {
    box-sizing: border-box;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 2px;
    border-radius: 50%;
    cursor: pointer;

    &:before {
        content: '';
        width: 100%;
        height: 100%;
        display: block;
        background-color: fade(@clearWhite, 50%);
        border-radius: 50%;
    }

    &_active {
        padding: 0;
        cursor: default;

        &:before {
            background-color: @clearWhite;
        }
    }

    &_active&_timer {
        &:before {
            display: none;
        }
    }

    &__timer {
        display: block;
        width: 100%;
        height: 100%;
        transform: rotate(90deg);
        border-radius: 50%;
    }

    &__timer-circle {
        fill: fade(@clearWhite, 50%);
        stroke: @clearWhite;
        stroke-width: 100;
        stroke-dasharray: 1000;
        animation: timer linear;
        transform: scale(-1, 1);
        transform-origin: center;
    }

    @keyframes timer {
        0% {
            stroke-dashoffset: -350;
        }

        100% {
            stroke-dashoffset: 0;
        }
    }
}
