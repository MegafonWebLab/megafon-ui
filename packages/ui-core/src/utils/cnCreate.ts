import classnames from 'classnames';

interface IModificators {
    [key: string]: boolean | string | undefined | null;
}

interface IClassSet {
    [key: string]: boolean;
}

type CustomClassNameType = string;
type CustomClassNamesArrayType = string[];
type CustomClassNamesType = CustomClassNameType | CustomClassNamesArrayType | undefined;

type ElementNameType = string | IModificators | CustomClassNamesArrayType | CustomClassNamesType;

const getCustomClassName = (first?: ElementNameType, second?: ElementNameType, third?: CustomClassNamesType) => {
    if (Array.isArray(first)) {
        return first.filter(f => typeof f === 'string').join(' ');
    }
    if (Array.isArray(second)) {
        return second.filter(f => typeof f === 'string').join(' ');
    }
    if (typeof second === 'string') {
        return second;
    }
    if (Array.isArray(third)) {
        return third.filter(f => typeof f === 'string').join(' ');
    }
    if (typeof third === 'string') {
        return third;
    }

    return '';
};

/**
 * Бэм генератор
 * @param {String} blockName - название блока
 * @returns {function} - возвращает функцию с тремя возожными параметрами
 * @param {String} elementName - название элемента
 * @param {Object} modificatorsObject - объект с модификаторами
 * @param {String} customClassNames - названия кастомных классов
 * @returns {String}
 */
function cnCreate(blockName: string) {
    function getElement(): string;
    function getElement(elementName: ElementNameType): string;
    function getElement(modificatorsObject: ElementNameType): string;
    function getElement(customClassNames: CustomClassNamesArrayType): string;
    function getElement(elementName: ElementNameType, modificatorsObject: ElementNameType): string;
    function getElement(elementName: ElementNameType, customClassNames: CustomClassNamesType): string;
    function getElement(modificatorsObject: ElementNameType, customClassNames: CustomClassNamesType): string;
    function getElement(
        elementName: ElementNameType,
        modificatorsObject: ElementNameType,
        customClassNames: CustomClassNamesType
    ): string;
    function getElement(
        elementName?: ElementNameType,
        modificatorsObject?: ElementNameType,
        customClassNames?: CustomClassNamesType
    ): string {
        const prefix = typeof elementName === 'string' ? elementName : '';
        const className = getCustomClassName(elementName, modificatorsObject, customClassNames);
        const secondMods = !Array.isArray(modificatorsObject) && typeof modificatorsObject !== 'string'
            ? modificatorsObject
            : undefined;
        const params = typeof elementName !== 'string' && !Array.isArray(elementName)
            ? elementName
            : secondMods;

        if (!params || typeof params === 'object' && Object.keys(params).length === 0) {
            return classnames(`${blockName}${prefix ? '__' + prefix : ''}`, className);
        }

        const classParams: IClassSet = {};
        let prefixKey = '', withoutPrefix = '';

        for (const key in params) {
            if (typeof params[key] === 'boolean' && params[key]) {
                prefixKey = `__${prefix}_${key}`;
                classParams[
                    `${blockName}${prefix ? prefixKey : '_' + key}`
                ] = true;
            } else if (typeof params[key] === 'string') {
                prefixKey = `__${prefix}_${key}_${params[key]}`;
                withoutPrefix = `_${key}_${params[key]}`;
                classParams[
                    `${blockName}${prefix ? prefixKey : withoutPrefix}`
                ] = true;
            }
        }

        return prefix === ''
            ? classnames(blockName, classParams, className)
            : classnames(`${blockName}__${prefix}`, classParams, className);
    }

    return getElement;
}

export default cnCreate;
