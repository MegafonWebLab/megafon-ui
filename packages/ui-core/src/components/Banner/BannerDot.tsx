import * as React from 'react';
import * as PropTypes from 'prop-types';
import './BannerDot.less';
import cnCreate from 'utils/cnCreate';

export interface IBannerDotProps {
    className?: string;
    index: number;
    isActive: boolean;
    showTimer: boolean;
    timerDelay: number;
    onClick: (index: number) => void;
}

const cn = cnCreate('mfui-beta-banner-dot');
const BannerDot: React.FC<IBannerDotProps> = ({
    className,
    index,
    isActive,
    showTimer,
    timerDelay,
    onClick,
}) => {
    const handleDotClick = React.useCallback(() => {
        onClick(index);
    }, [onClick, index]);

    return (
        <div
            className={cn({ active: isActive, timer: showTimer }, className)}
            onClick={handleDotClick}
        >
            {showTimer && isActive && (
                <svg className={cn('timer')} viewBox="0 0 100 100">
                    <circle
                        className={cn('timer-circle')}
                        style={{ animationDuration: `${timerDelay}s` }}
                        cx="50"
                        cy="50"
                        r="50"
                    />
                </svg>
            )}
        </div>
    );
};

BannerDot.propTypes = {
    className: PropTypes.string,
    index: PropTypes.number.isRequired,
    isActive: PropTypes.bool.isRequired,
    showTimer: PropTypes.bool.isRequired,
    timerDelay: PropTypes.number.isRequired,
    onClick: PropTypes.func.isRequired,
};

export default BannerDot;
