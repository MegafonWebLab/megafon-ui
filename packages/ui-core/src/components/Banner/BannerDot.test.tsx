import * as React from 'react';
import { shallow } from 'enzyme';
import cnCreate from '../../utils/cnCreate';
import BannerDot, { IBannerDotProps } from './BannerDot';

const props = {
    index: 3,
    isActive: false,
    showTimer: false,
    timerDelay: 0,
    onClick: jest.fn(),
} as IBannerDotProps;

const cn = cnCreate('.mfui-beta-banner-dot');

describe('<BannerDot />', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render with props', () => {
        const wrapper = shallow(
            <BannerDot {...props} />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render with className', () => {
        const wrapper = shallow(
            <BannerDot {...props} className="class" />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render without timer when isActive is true and showTimer is false', () => {
        const wrapper = shallow(
            <BannerDot {...props} isActive />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render without timer when isActive is false and showTimer is true', () => {
        const wrapper = shallow(
            <BannerDot {...props} showTimer />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render timer when isActive is true and showTimer is true', () => {
        const wrapper = shallow(
            <BannerDot {...props} showTimer isActive />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should call onClick', () => {
        const wrapper = shallow(
            <BannerDot {...props} showTimer isActive />
        );

        wrapper.find(cn()).simulate('click');

        expect(props.onClick).toBeCalledWith(props.index);
    });
});
