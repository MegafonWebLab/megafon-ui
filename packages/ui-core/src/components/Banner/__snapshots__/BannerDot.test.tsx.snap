// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BannerDot /> should render timer when isActive is true and showTimer is true 1`] = `
<div
  className="mfui-beta-banner-dot mfui-beta-banner-dot_active mfui-beta-banner-dot_timer"
  onClick={[Function]}
>
  <svg
    className="mfui-beta-banner-dot__timer"
    viewBox="0 0 100 100"
  >
    <circle
      className="mfui-beta-banner-dot__timer-circle"
      cx="50"
      cy="50"
      r="50"
      style={
        Object {
          "animationDuration": "0s",
        }
      }
    />
  </svg>
</div>
`;

exports[`<BannerDot /> should render with className 1`] = `
<div
  className="mfui-beta-banner-dot class"
  onClick={[Function]}
/>
`;

exports[`<BannerDot /> should render with props 1`] = `
<div
  className="mfui-beta-banner-dot"
  onClick={[Function]}
/>
`;

exports[`<BannerDot /> should render without timer when isActive is false and showTimer is true 1`] = `
<div
  className="mfui-beta-banner-dot mfui-beta-banner-dot_timer"
  onClick={[Function]}
/>
`;

exports[`<BannerDot /> should render without timer when isActive is true and showTimer is false 1`] = `
<div
  className="mfui-beta-banner-dot mfui-beta-banner-dot_active"
  onClick={[Function]}
/>
`;
